src/hp.rs

```diff
@@ -34,11 +35,28 @@ impl Hp {
         self.state
     }
 
-    pub fn damage(&mut self, damage: u16) {
-        self.current -= damage;
-        if self.current == 0 {
-            self.state = HpStateEnum::Dying;
-        }
+    pub fn heal(&mut self, value: u16) {
+        self.current = min(
+            match self.current().checked_add(value) {
+                Some(val) => val,
+                None => u16::MAX
+            },
+            self.total()
+        )
+
+    }
+
+    pub fn damage(&mut self, value: u16) {
+        self.current = max(
+            match self.current().checked_sub(value) {
+                Some(val) => val,
+                None => u16::MIN
+            },
+            0
+        );
+        if self.current() == 0 {
+            self.state = HpStateEnum::Dying
+        };
     }
```

Comment:
Oh, I figured it out. This could probably be changed as follows if you'd like:
```suggestion
    pub fn damage(&mut self, value: u16) {
        self.current = match self.current().checked_sub(value) {
                Some(val) => val,
                None => u16::MIN
            };
        if self.current() == 0 {
            self.state = HpStateEnum::Dying
        };
    }
```

Reply:
I messed up and committed a change instead of suggesting something. Have a look below at the latest commit please.